### PR TITLE
Sort career listings by location in the CMS

### PR DESCRIPTION
**Pull request by andrewjdick**
*****
Do not merge until `backend: branch` has been reverted to `master`
*****

## Ticket

_Which GitHub issue does your code resolve?_

Closes: #438 
Closes: #445 

## Proposed Changes

_What changes does your code introduce?_

- ✨ Sorts CMS career listings by location.
- ✨ Adds location and duration attributes to career slugs.
- ✨ Adds default location option when creating a new career in the cms. 
- 🎨 Tidied up career file pathnames

## Types of changes

_What types of changes does your code introduce?_

- [ ] 🔥 Breaking change (fix or feature that would cause existing functionality to change)
- [x] ✨ New feature (non-breaking change which adds functionality)
- [ ] 🐛 Bug fix (non-breaking change which fixes an issue)
- [x] 🎨 Improve general format/structure
- [ ] ⬆️ Upgrading dependencies
- [ ] 🔧 Configuration

## Screenshots / Screen Recordings

**Careers sorted by location:** 
image

**New career default location:**
defaultlocation

**Same position being used in different locations:**
location-position

### CODE REVIEW

File: static/admin/config.yml [hunk 1]
```diff
@@ -1,6 +1,6 @@
 backend:
   name: git-gateway
-  branch: master
+  branch: career-collection-ny
```
Comment (andrewjdick) [inline static/admin/config.yml]: TODO: Revert this to master when we are happy with the implementation

File: gatsby/onCreateNode.js [hunk 2]
```diff
@@ -3,19 +3,24 @@ var slugify = require('slugify');
...
-			});
+			const { title, location, duration } = frontmatter;
+
+			const formattedLocation = slugify(location, slugFormatOptions);
+			const formattedDuration = slugify(duration, slugFormatOptions);
+			const formattedTitle = slugify(title, slugFormatOptions);
 
 			// This string will be used as the final URL
-			const slug = `/${formattedTitle}.html`;
+			const slug = `/${formattedLocation}/${formattedDuration}/${formattedTitle}.html`;
```
Comment (ibrennan) [inline gatsby/onCreateNode.js]: This approach can still lead to a clash, I'd rather see something like:

```
const hash = Math.random().toString(36).substring(7);
const slug = `/${hash}/${formattedLocation}/${formattedTitle}.html`;
```

I like having the location and title in the URL, so left that bit in. But added a snippet to generate and assign a random hash to the URL to always have a unique slug regardless.

If you want to be a bit more verbose feel free to create a hash generator in a utils file, or get one from npm.

File: static/admin/config.yml [hunk 3]
```diff
@@ -69,7 +71,43 @@ collections:
           label: 'Location',
           name: 'location',
           widget: 'select',
-          options: ['London', 'New York', 'Singapore', 'Lisbon'],
+          options: ['London', 'New York'],
```
Comment (ibrennan) [inline static/admin/config.yml]: Technically `New York City` if we are going by city names, can you update this across the place?

File: gatsby/onCreateNode.js [hunk 4]
```diff
@@ -38,10 +43,7 @@ module.exports = ({ node, actions, getNode }) => {
 			const { date, title, hero } = frontmatter;
 
 			const formattedDate = date.split('-').join('/');
-			const formattedTitle = slugify(title, {
-				lower: true,
-				remove: /[$*_+~.()'"!?\-:@]/g,
-			});
+			const formattedTitle = slugify(title, slugFormatOptions);
```
Comment (jjenzz) [inline gatsby/onCreateNode.js]: Do we need to do this twice? Seems we can move this declaration below `const slugFormatOptions` and reuse it. I won't hold up the PR for this though.

File: gatsby/onCreateNode.js (hunk 2 above)
Comment (jjenzz) [reply gatsby/onCreateNode.js]: @andrewjdick I added the `nanoid` package to the repo when I did the careers tabs. You might be able to use that for this.

File: static/admin/config.yml (hunk 3 above)
Comment (ibrennan) [reply static/admin/config.yml]: Actually, ignore this feedback. I did a bit of research on competitors and they seem to all use `New York` when describing the city, and `NY` when describing the state.

Comment (andrewjdick) [reply static/admin/config.yml]: Haha cool

**Comment by ibrennan** [comment]
Notes from discussion:

URL for a career detail item should be:

`/careers/${hash}-${formattedLocation}-${formattedDuration}-${formattedTitle}.html`;

File: gatsby/onCreateNode.js (hunk 4 above)
Comment (andrewjdick) [reply gatsby/onCreateNode.js]: Good spot. Using `formatField()` to generate the formatted attributes

File: src/pages/careers.js [hunk 5]
```diff
@@ -1,4 +1,4 @@
-import React, { Fragment } from 'react';
+import React from 'react';
```
Comment (andrewjdick) [inline src/pages/careers.js]: Unused Fragment

File: gatsby/onCreateNode.js (hunk 2 above)
Comment (andrewjdick) [reply gatsby/onCreateNode.js]: Discussion notes:

- We can't use `nanoid` in `config.yml` to generate a unique hash for the filename.
- Shelving for now.

**Comment by andrewjdick** [comment]
@ibrennan @peduarte requested changes in 👍 

wallop

File: gatsby/onCreateNode.js (hunk 4 above)
Comment (jjenzz) [reply gatsby/onCreateNode.js]: Ooops, sorry, that's not what I was referring to. The line I commented on was line `46` which has `formattedTitle` declaration when you have already declared a `formattedTitle` on line `21`. I was suggesting having just one `formattedTitle` declaration below `const slugFormatOptions`. The `slugFormatOptions` approach was fine as it was.

File: gatsby/onCreateNode.js [hunk 6]
```diff
@@ -3,19 +3,25 @@ var slugify = require('slugify');
...
-			});
+			const { title, location, duration } = frontmatter;
+
+			const formattedLocation = slugifyField(location);
+			const formattedDuration = slugifyField(duration);
+			const formattedTitle = slugifyField(title);
 
 			// This string will be used as the final URL
-			const slug = `/${formattedTitle}.html`;
+			const slug = `/${formattedLocation}-${formattedDuration}-${formattedTitle}.html`;
```
Comment (ibrennan) [inline gatsby/onCreateNode.js]: `/careers` here at the start of the URL

Comment (andrewjdick) [reply gatsby/onCreateNode.js]: @ibrennan done 👍

**Pull request merged**